### PR TITLE
Map tip preview

### PR DESCRIPTION
**Pull request by YoannQDQ**
## Description

Adds a map tip preview in the Raster and Vector layer properties dialog.
This makes it easier to design map tips.

map_tip_preview

### CODE REVIEW

File: src/gui/raster/qgsrasterlayerproperties.h [hunk 1]
```diff
@@ -305,5 +313,11 @@ class GUI_EXPORT QgsRasterLayerProperties : public QgsOptionsDialogBase, private
     QgsCoordinateReferenceSystem mBackupCrs;
 
     QgsRasterAttributeTableWidget *mRasterAttributeTableWidget = nullptr;
+
+    void initMapTipPreview();
+
+    QWidget *mMapTipPreviewWidget = nullptr;
+    QgsWebView *mMapTipPreview = nullptr;
+    const int MARGIN_VALUE = 5;
```
Comment (elpaso) [inline src/gui/raster/qgsrasterlayerproperties.h:321]: Maybe static const and initialize in cpp file.

File: src/gui/qgsmaptip.cpp [hunk 2]
```diff
@@ -373,3 +379,71 @@ void QgsMapTip::onLinkClicked( const QUrl &url )
...
+
+  // Insert the map tip text into the html template
+  return QgsMapTip::htmlText( tipText );
+
+}
+
+QString QgsMapTip::rasterMapTipPreviewText( QgsMapLayer *layer, QgsMapCanvas *mapCanvas, const QString &mapTemplate )
+{
+  QgsRasterLayer *rlayer = qobject_cast<QgsRasterLayer *>( layer );
+  if ( !rlayer || mapTemplate.isEmpty() )
```
Comment (elpaso) [inline src/gui/qgsmaptip.cpp:433]: Can be safer to check for mapCanvas
```suggestion
  if (!mapCanvas || !rlayer || mapTemplate.isEmpty() )
```

File: src/gui/qgsmaptip.cpp [hunk 3]
```diff
@@ -373,3 +379,71 @@ void QgsMapTip::onLinkClicked( const QUrl &url )
 {
   QDesktopServices::openUrl( url );
 }
+
+
+QString QgsMapTip::vectorMapTipPreviewText( QgsMapLayer *layer, QgsMapCanvas *mapCanvas, const QString &mapTemplate, const QString &displayExpression )
+{
+  // Only spatial layers can have map tips
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( layer );
+  if ( !vlayer || !vlayer->isSpatial() )
```
Comment (elpaso) [inline src/gui/qgsmaptip.cpp:388]:
```suggestion
  if (!mapCanvas || !vlayer || !vlayer->isSpatial() )
```

**Review by elpaso — COMMENTED** [review]
Very nice work!

File: src/gui/qgsmaptip.cpp (hunk 3 above)
Comment (YoannQDQ) [reply src/gui/qgsmaptip.cpp:388]: Ok

File: src/gui/qgsmaptip.cpp (hunk 2 above)
Comment (YoannQDQ) [reply src/gui/qgsmaptip.cpp:433]: Ok

File: src/gui/qgsmaptip.cpp [hunk 4]
```diff
@@ -336,28 +320,50 @@ QString QgsMapTip::fetchRaster( QgsMapLayer *layer, QgsPointXY &mapPosition, Qgs
 
   if ( !layer->isInScaleRange( mapCanvas->mapSettings().scale() ) ||
        ( mapCanvas->mapSettings().isTemporal() && !layer->temporalProperties()->isVisibleInTemporalRange( mapCanvas->temporalRange() ) ) )
-  {
     return QString();
-  }
+
+  if ( rlayer->mapTipTemplate().isEmpty() )
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsmaptip.cpp:325]: Ok

File: src/gui/raster/qgsrasterlayerproperties.h (hunk 1 above)
Comment (YoannQDQ) [reply src/gui/raster/qgsrasterlayerproperties.h:321]: I can make it static, but I believe most` static const int` in the code base are initialized in the header. Anyway, if this PR is accepted, I intend to propose a new PR that makes the margin, border and background color configurable, and make it possible to have semi-transparent map tips, so MARGIN_VALUE would disappear.

**Review by elpaso — APPROVED** [review]

**Pull request merged**